File: Minimum Swaps to Group All 1's Together II.cpp
class Solution {
public:
    int minSwaps(vector<int>& nums) {
        int k = reduce(nums.begin(), nums.end());
        if (k == 0) return 0;

        int n = nums.size();
        int total = reduce(nums.begin(), nums.begin() + k-1);
        int i = 0, j = k-1;
        int maxx = 0;
        while (i < n) {
            total += nums[j++];
            j %= n;
            maxx = max(maxx, total);
            total -= nums[i++];
        }
        return k - maxx;
    }
};

File: Minimum Swaps to Group All 1's Together II.py
class Solution:
    def minSwaps(self, nums: List[int]) -> int:
        k = sum(nums)
        n = len(nums)
        total = 0
        for i in range(k - 1):
            total += nums[i]
        i = 0
        j = k - 1
        maxx = 0
        while i < n:
            total += nums[j]
            j += 1
            j %= n
            maxx = max(maxx, total)
            total -= nums[i]
            i += 1
        return k - maxx
